### ISSUE TITLE
ng build stuck after successful build since sentryWebpack v2  

### ISSUE THREAD

**Issue by johanneszelger** [getsentry/sentry-javascript-bundler-plugins#345]
Since i upgraded from the plugin v1 to v2, my angular build gets stuck after a successfull build:

`Build at: 2023-07-07T10:40:52.046Z - Hash: cc2971c88cd12369 - Time: 32093ms`

just won't end the process after this line. When i remove the plugin, the build ends normally.  I tried the plugin versions 2.4 2.3 2.0, all unsuccessfull.

**Comment by Lms24** [comment]
Hi, can you please share your plugin config? Do you get console output from the plugin regarding source maps upload or how many files were uploaded?

**Comment by johanneszelger** [comment]
Sure, my config is:

`plugins = [
  new webpack.DefinePlugin({
    "process.env.RELEASE": JSON.stringify(release),
    "process.env.BUILD": JSON.stringify(build)
  })
]

if (process.env.SENTRY_AUTH_TOKEN !== undefined && process.env.SENTRY_AUTH_TOKEN !== "") {
  console.log("adding sentry webpack because auth token was found, release: ", release)
  plugins.push(sentryWebpackPlugin({
    url: "url",
    org: "org",
    project: "frontend",

    // Auth tokens can be obtained from https://sentry.io/settings/account/api/auth-tokens/
    // and need the `project:releases` and `org:read` scopes
    authToken: process.env.SENTRY_AUTH_TOKEN,

    // Specify the directory containing build artifacts
    // By default, Angular emits build files in the `dist` directory but your
    // configuration may vary
    // sourcemaps: {
    //   assets: ["./dist/Incentrium"]
    // },

    // Optionally uncomment the line below to override automatic release name detection
    release: {
      name: release,
      finalize: false,
      cleanArtifacts: true,
      dist: build
    }
  }))
} else
  console.log("skipping sentry webpack because auth token was not found")

module.exports = {
  devtool: "source-map", // Source map generation must be turned on
  plugins: plugins,
  watch: false,
};`

And yes i get output from the plugin, all sourcemaps are uploaded successfully, they are also visible on sentry. Everything is working - its just that my build pipeline gets stuck after the successfull build. If i cancel the process manually, everything works as expected.

**Comment by MitkoTschimev** [comment]
We have the same issue. It just doesn't end the webpack process

**Comment by MitkoTschimev** [comment]
I was able to pin it down to the dependency unplugin. 
For a simple plugin with this dep on the newest version shows the same behavior

**Comment by Lms24** [comment]
Thanks for investigating this @MitkoTschimev, appreciate the effort! Let's see if this can be fixed in unplugin. In the meantime, you might want to look into setting up [Sentry-CLI](https://docs.sentry.io/platforms/javascript/guides/angular/sourcemaps/uploading/cli/) to upload your source maps as a workaround.

**Comment by MitkoTschimev** [comment]
It is unsuitable for us because we use service workers with angular. The service worker works based on the content hash value of artifacts. if we manipulate afterwords the files, they are not matching anymore.

This is the advantage of the webpack plugin.

**Comment by lforst** [comment]
I am contemplating whether we should just remove unplugin entirely. In the beginning, I thought it was a good idea to abstract away all the bundler intricacies but by now many of the injection flows are different anyhow. I believe we're operating too low-level for unplugin to be effective. It would be a medium refactor but one dependency less.

**Comment by johanneszelger** [comment]
> I am contemplating whether we should just remove unplugin entirely. In the beginning, I thought it was a good idea to abstract away all the bundler intricacies but by now many of the injection flows are different anyhow. I believe we're operating too low-level for unplugin to be effective. It would be a medium refactor but one dependency less.

Is there any news yet?

**Comment by niewicz** [comment]
Any updates?

**Comment by lforst** [comment]
No updates. I think we are soft-blocked by https://github.com/unjs/unplugin/issues/323. We could remove the `unplugin` dependency but it would result in quite a bit of work which we don't have the cycles for right now :(

**Comment by krema** [comment]
Any updates?

**Comment by rgant** [comment]
The documentation currently says to do what doesn't work. https://docs.sentry.io/platforms/javascript/guides/angular/sourcemaps/uploading/angular-webpack/

If there isn't bandwidth to remove unplugin, perhaps there is bandwidth to update the documentation with an alternative setup?

Because otherwise I will continue to use `"@sentry/webpack-plugin": "^1.21.0",` until this is resolved.

**Comment by lforst** [comment]
@rgant Using `"@sentry/webpack-plugin": "^1.21.0"` is perfectly fine!

**Comment by Agrumas** [comment]
Any news? This occurs to us too, however downgrading to `^1.21.0` isn't an option

**Comment by lforst** [comment]
@Agrumas No news.

**Comment by salonmonster** [comment]
This is still an issue.
I've rolled back to ^1.21.0 for now, but wasted a good while trying to track this down :(

**Comment by gerbermichi** [comment]
Our workaround is to use the sentry-cli to upload the sourcemap with a specific release:
[Associating release with Artifact Bundle](https://docs.sentry.io/platforms/javascript/sourcemaps/uploading/cli/#associating-release-with-artifact-bundle)

**Comment by andreiborza** [comment]
Yes, so the workaround is either to downgrade or to use sentry-cli. We are thinking of refactoring unplugin out but have not yet committed to it.

**Comment by smeubank** [comment]
> Let's see if this can be fixed in unplugin

@Lms24 any update if this is fixable in unplugin? This came up internally from another org. Realistically i do not think we will remove unplugin in the near future. Is there an existing issue here https://github.com/unjs/unplugin/issues ? If not we can create one

**Comment by Lms24** [comment]
Hmm no updates from my end but the suggested fix sounds simple enough to give it a try! We can try first in our plugin (will do some testing) and if this works, we could submit a PR upstream.

**Comment by Lms24** [comment]
I opened https://github.com/getsentry/sentry-javascript-bundler-plugins/pull/653 but no promises that this will get merged or work. As it stands, I actually cannot reproduce this issue locally in a simple Angular 14 test app that uses the Angular webpack builder and our plugin in the latest version.

If someone can provide a minimal reproduction (zip, GH issue, whatever else but please buildable project), I'd appreciate it!

**Comment by Lms24** [comment]
Something else I just came across: https://stackoverflow.com/a/63246847. Might be worth for anyone experiencing this issue to set `watch` explicitly. Again, I cannot reproduce this, so I can't help any further.

**Comment by rgant** [comment]
Angular v14 is _really_ old and [well past end of support](https://angular.dev/reference/releases#actively-supported-versions). IIRC this was a problem with updating to v16 (based on the age of my original comment), and definately an issue with v17 (which is the oldest version still officially supported).

Also note that that SO post is also quite old and ejecting the Angular webpack config is no longer event a thing I believe. Not even sure if webpack is still much of a thing anymore: https://angular.dev/tools/cli/build-system-migration

**Comment by Lms24** [comment]
I'm well aware of Angular 14 being old :) However, https://github.com/unjs/unplugin/issues/323#issuecomment-2020010092  mentions NG14 multiple times, so I thought I'd try to repro the bug on 14. Also aware of Webpack not being used anymore but afaik, you can still opt into using the webpack builder. Which anyone who's still coming across this problem must be using anyway :)

**Comment by Lms24** [comment]
We released version 3.0.0 of the plugins. It includes a potential fix for the Angular build hanging (https://github.com/getsentry/sentry-javascript-bundler-plugins/pull/653). Would be great if anyone still affected by this can give it a try and report back. Thanks!

**Comment by Lms24** [comment]
Quick update: We had to change course a bit and not the forced process exit that I introduced in #653 no longer is default behaviour but since version `3.1.0` you need to enable it with an experimental flag (see #663):

```js
sentryWebpackPlugin({
  // ...other options
  _experiments: { forceExitOnBuildCompletion: true },
});
```
